src/varity/ref_gene.clj

```diff
@@ -155,6 +156,102 @@
 
+;;;; genomic.gff
+(defn- parse-genomic-gff-line
+  [s & {:keys [attr-kv-sep] :or {attr-kv-sep (gtf-gff-attr-kv-sep :gff3)}}]
+  (when-not (= \# (first s))
+    (let [row (string/split s #"\t")]
+      (merge
+       {:seqid (nth row 0)
+        :source (nth row 1)
+        :type (nth row 2)
+        :start (as-long (nth row 3))
+        :end (as-long (nth row 4))
+        :score (as-float (nth row 5))
+        :strand (nth row 6)
+        :phase (nth row 7)}
+       (when (= (count row) 9)
+         {:attributes (->gtf-gff-attr (nth row 8) attr-kv-sep)})))))
+
+(def ^:private nm-pattern #"NM_\d+\.\d+")
+(def ^:private nr-pattern #"NR_\d+\.\d+")
+
+(def find-nm-id (memoize #(re-find nm-pattern %)))
+(def find-nr-id (memoize #(re-find nr-pattern %)))
+
+(def chr-idx-name-map (merge (zipmap (range 1 23) (map str (range 1 23))) {23 "X" 24 "Y"}))
+
+(defn get-chr-from-seqid
+  [seqid]
+  (let [accession-number (re-find #"NC_0000(\d+)" seqid)
+        chr-idx (when accession-number
+                  (Integer/parseInt (second accession-number)))
+        chr-name (chr-idx-name-map chr-idx)]
+    (when chr-name
+      (str "chr" chr-name))))
+
+(defn get-gene-id
+  [dbxref]
+  (when-let [gene-id (re-find #"GeneID:(\d+)" dbxref)]
+    (second gene-id)))
+
+(defn find-nm-from-attributes
+  [attributes]
+  (let [id (find-nm-id (attributes "ID"))
+        parent (find-nm-id (get attributes "Parent" ""))]
+    (or id parent)))
+
+(defn find-nr-from-attributes
+  [attributes]
+  (let [id (find-nr-id (attributes "ID"))]
+    id))
+
+(defn get-exon-number-from-id
+  [id]
+  (nth (re-find #"exon-(NM|NR)_\d+.\d+-(\d+)" id) 2))
+
+(defn- build-genomic-gff-feature-map
+  ([features]
+   (build-genomic-gff-feature-map features (zipmap [:transcript :exon :cds] (repeat {}))))
+  ([features data]
+   (reduce (fn [data {:keys [start end seqid type attributes] :as gff}]
+             (let [base {:start start
+                         :end end
+                         :chr (get-chr-from-seqid seqid)}]
+               (if-let [t-id (or (find-nm-from-attributes attributes)
+                                 (find-nr-from-attributes attributes))]
+                 (case type
+                   ("mRNA" "transcript" "lnc_RNA")
+                   (assoc-in data
+                             [:transcript t-id]
+                             (merge base
+                                    {:name t-id
+                                     :name2 (get attributes "gene")
+                                     :gene-id (get-gene-id (get attributes "Dbxref"))
+                                     :strand (:strand gff)
+                                     :score (:score gff)}))
+
+                   "exon"
+                   (update-in data
+                              [:exon t-id]
+                              conj
+                              (merge base
+                                     {:exon-number (as-long (get-exon-number-from-id (attributes "ID")))
```

Comment:
*_genomic.gff file doesn’t have exon_number in attributes, so I just get exon number from ID.

And considering that “ID” is unique identifier, I should get accession number from transcript_id or Parent in attributes.

I will fix them. Thank you.

Reply:
> *_genomic.gff file doesn’t have exon_number in attributes, so I just get exon number from ID.

It does but I found only 3 rows in a file 😕 
I found genomic.gtf.gz contains exon_number and some other useful data (e.g. [this version](https://ftp.ncbi.nlm.nih.gov/genomes/refseq/vertebrate_mammalian/Homo_sapiens/all_assembly_versions/GCF_000001405.25_GRCh37.p13/)). Is there any reason to prefer gff over gtf? I think gtf is more convenient.

> I should get accession number from transcript_id or Parent in attributes.

The doc says Parent refers to ID so I doubt we can use that. 
It guess we need to use the key `protein_id` for NP_,  `transcript_id` for NM_ and order of exon records for exon-number (in case of gff).

> exon_number
>   The GFF3 specification does not include explicit numbering of exons, as is
>   sometimes found in GTF files. Occasionally an exon_number attribute is 
>   present on exon features that are annotated separately from an RNA feature. 
>   Exon features that are children of mRNA or other transcript features do not
>   include an explicit exon_number attribute, and their relative order should be
>   inferred from the order on the genome.

Reply:
I fail to notice the .gtf file and I’ll check it.
And I should consider attributes and exon-number.
Thank you for your comments!